bots/mainbot/movement.js

```diff
@@ -124,39 +134,89 @@ movement.selectNext = (openList) => {
 };
 
 movement.getOpenAdj = (map, curr) => {
-
-    // All adjacent nodes
-    let adj;
-    if(!map[0][curr.x+1]){
-        adj = [map[curr.y][curr.x - 1], map[curr.y + 1][curr.x], map[curr.y - 1][curr.x],
-               map[curr.y - 1][curr.x - 1], map[curr.y + 1][curr.x -1]];
+    if(!map || !curr || !map[curr.y] || !map[0][curr.x]){
+        return [];
     }
-    else if(!map[0][curr.x-1]){
-        adj = [map[curr.y][curr.x + 1], map[curr.y + 1][curr.x], map[curr.y - 1][curr.x],
-               map[curr.y + 1][curr.x + 1], map[curr.y - 1][curr.x + 1]];
+
+    let y = curr.y;
+    let x = curr.x;
+    let y_length = map.length -1;
+    let x_length = map[0].length -1;
+
+    let adj = [];
+
+    if(y < y_length){
+        adj.push(map[y+1][x]);
+        // If in x bounds
+        if(x < x_length){
+            adj.push(map[y+1][x+1])
+        }
+        if(x > 0){
+            adj.push(map[y+1][x-1])
+        }
     }
-    else if(!map[curr.y+1]){
-        adj = [map[curr.y][curr.x + 1], map[curr.y][curr.x - 1], map[curr.y - 1][curr.x],
-               map[curr.y - 1][curr.x - 1], map[curr.y - 1][curr.x + 1]];
+    if(y > 0){
+        adj.push(map[y-1][x]);
+        if(x < x_length){
+            adj.push(map[y-1][x+1])
+        }
+        if(x > 0){
+            adj.push(map[y-1][x-1])
+        }
+
+
     }
-    else if(!map[curr.y-1]){
-        adj = [map[curr.y][curr.x + 1], map[curr.y][curr.x - 1], map[curr.y + 1][curr.x],
-               map[curr.y + 1][curr.x + 1], map[curr.y + 1][curr.x -1]];
+    if(x > 0){
+        adj.push(map[y][x-1])
     }
-    else{
-        adj = [map[curr.y][curr.x + 1], map[curr.y][curr.x - 1], map[curr.y + 1][curr.x], map[curr.y - 1][curr.x],
-               map[curr.y + 1][curr.x + 1], map[curr.y - 1][curr.x - 1], map[curr.y - 1][curr.x + 1], map[curr.y + 1][curr.x -1]];
+    if(x < x_length){
+        adj.push(map[y][x+1])
     }
+
     // All adjacent nodes that are open
     let openAdj = [];
 
-    for(let i = 0; i < adj.length; ++i){
+    for(let i = 0; i < adj.length; i+=1){
         if(adj[i] && !adj[i].closed){
             openAdj.push(adj[i])
         }
     }
     return openAdj
 };
 
+movement.condense_path = (speed, path) => {
+    let step_start = 0;
+    let i = 0;
+    let condensed_path = [];
+
+    if(!path){
+        return [];
+    }
+
+    while(step_start < path.length -1){
+        // Loop while the bot can reach that square
+        for(i = step_start + 1; movement.get_dist(path[step_start], path[i]) < speed; i += 1){}
+
+        if(i == path.length-1 && movement.get_dist(path[step_start], path[i]) < speed){
+            condensed_path.push(path[i])
+            step_start = i;
+        }
+        else{
+            condensed_path.push(path[i-1]);
+            step_start = i-1;
+        }
+
+    }
+
+    return condensed_path;
+};
+
+movement.get_dist = (start, end) => {
```

Comment:
This function already exists in utilities

Reply:
I am using two different distance types. The one in utilities is manhattan distance which is just dx + dy kind of thing and is used in the A* algorithm. The other is the distance formula used the check how far a node is for robot movement.

Reply:
I will move them both into utils file and rename them